FDRS_Gateway/fdrs_functions.h

```diff
@@ -191,37 +196,46 @@ void getSerial() {
 
   }
 }
-void sendSD(const char filename[32]) {
-#ifdef USE_SD_LOG
-  DBG("Logging to SD card.");
-  File logfile = SD.open(filename, FILE_WRITE);
-  for (int i = 0; i < ln; i++) {
-    char linebuf[32];
-#ifdef USE_WIFI
-    sprintf(linebuf, "%ld,%d,%d,%g", timeClient.getEpochTime(), theData[i].id, theData[i].t, theData[i].d);
-#else
-    sprintf(linebuf, "%ld,%d,%d,%g", seconds_since_reset, theData[i].id, theData[i].t, theData[i].d);
-#endif
-    logfile.println(linebuf);
-  }
+#if defined (USE_SD_LOG) || defined (USE_FS_LOG)
+void releaseLogBuffer()
+{
+  #ifdef USE_SD_LOG
+  DBG("Releasing Log buffer to SD");
+  File logfile = SD.open(SD_FILENAME, FILE_WRITE);
+  logfile.print(logBuffer);
   logfile.close();
-#endif
+  #endif
+  #ifdef USE_FS_LOG
+  DBG("Releasing Log buffer to internal flash.");
+  File logfile = LittleFS.open(FS_FILENAME, "a");
+  logfile.print(logBuffer);
+  logfile.close();
+  #endif
+  memset(&(logBuffer[0]), 0, sizeof(logBuffer)/sizeof(char));
+  logBufferPos = 0;
 }
-void sendFS(const char filename[32]) {
-#ifdef USE_FS_LOG
-  DBG("Logging to internal flash.");
-  File logfile = LittleFS.open(filename, "a");
-  for (int i = 0; i < ln; i++) {
-    char linebuf[32];
-#ifdef USE_WIFI
-    sprintf(linebuf, "%ld,%d,%d,%g", timeClient.getEpochTime(), theData[i].id, theData[i].t, theData[i].d);
-#else
-    sprintf(linebuf, "%ld,%d,%d,%g", seconds_since_reset, theData[i].id, theData[i].t, theData[i].d);
 #endif
-    logfile.println(linebuf);
+void sendLog()
+{
+  #if defined (USE_SD_LOG) || defined (USE_FS_LOG)
+  DBG("Logging to buffer");
+  for (int i = 0; i < ln; i++)
+  {
+    char linebuf[34]; // size depends on resulting length of the formatting string
+    #ifdef USE_WIFI
+    sprintf(linebuf, "%ld,%d,%d,%g\r\n", timeClient.getEpochTime(), theData[i].id, theData[i].t, theData[i].d);
+    #else
+    sprintf(linebuf, "%ld,%d,%d,%g\r\n", seconds_since_reset, theData[i].id, theData[i].t, theData[i].d);
+    #endif
+
+    if (logBufferPos+strlen(linebuf) >= (sizeof(logBuffer)/sizeof(char))) // if buffer would overflow, release first
+    {
+      releaseLogBuffer();
+    }
```

Comment:
Is it only released when the buffer is full? Could there be unreleased data in the buffer when MQTT reconnects that never gets out?

Reply:
Atm, this is the case. I was thinking of also releasing the buffer regularly in the loop when it has something in it.